src/coprocessor/dag/batch/executors/simple_aggr_executor.rs

```diff
@@ -0,0 +1,874 @@
+                    "Unable to use BatchSimpleAggregateExecutor: {}",
+                    e
+                ))
+            })?;
+        }
+        Ok(())
+    }
+}
+
+impl<C: ExecSummaryCollector, Src: BatchExecutor> BatchSimpleAggregationExecutor<C, Src> {
+    pub fn new(
+        summary_collector: C,
+        config: Arc<EvalConfig>,
+        src: Src,
+        aggr_definitions: Vec<Expr>,
+    ) -> Result<Self> {
+        Self::new_impl(
+            summary_collector,
+            config,
+            src,
+            aggr_definitions,
+            AggrDefinitionParser::parse,
+        )
+    }
+
+    /// Provides ability to customize `AggrDefinitionParser::parse`. Useful in tests.
+    fn new_impl<F>(
+        summary_collector: C,
+        config: Arc<EvalConfig>,
+        src: Src,
+        aggr_definitions: Vec<Expr>,
+        parse_aggr_definition: F,
+    ) -> Result<Self>
+    where
+        F: Fn(
+            Expr,
+            &Tz,
+            usize,
+            &mut Vec<FieldType>,
+            &mut Vec<RpnExpression>,
+        ) -> Result<Box<dyn AggrFunction>>,
+    {
+        let aggr_len = aggr_definitions.len();
+        if aggr_len == 0 {
+            return Err(box_err!("There should be at least one aggregate function"));
+        }
+
+        let mut aggr_fn_states = Vec::with_capacity(aggr_len);
+        let mut aggr_fn_output_cardinality = Vec::with_capacity(aggr_len);
+        let mut ordered_schema = Vec::with_capacity(aggr_len * 2);
+        let mut ordered_aggr_fn_input_exprs = Vec::with_capacity(aggr_len * 2);
+
+        let schema = src.schema();
+        let schema_len = schema.len();
+
+        for def in aggr_definitions {
+            let aggr_output_len = ordered_schema.len();
+            let aggr_input_len = ordered_aggr_fn_input_exprs.len();
+
+            let aggr_fn = parse_aggr_definition(
+                def,
+                &config.tz,
+                schema_len,
+                &mut ordered_schema,
+                &mut ordered_aggr_fn_input_exprs,
+            )?;
+
+            assert!(ordered_schema.len() > aggr_output_len);
+            let this_aggr_output_len = ordered_schema.len() - aggr_output_len;
+
+            // We only support 1 parameter aggregate functions, so let's simply assert it.
+            assert_eq!(ordered_aggr_fn_input_exprs.len(), aggr_input_len + 1);
+
+            aggr_fn_states.push(aggr_fn.create_state());
+            aggr_fn_output_cardinality.push(this_aggr_output_len);
+        }
+
+        let ordered_aggr_fn_output_types = ordered_schema
+            .iter()
+            .map(|ft| {
+                // The unwrap is fine because aggregate function parser should never return an
+                // eval type that we cannot process later. If we made a mistake there, then we
+                // should panic.
+                EvalType::try_from(ft.tp()).unwrap()
+            })
+            .collect();
+
+        let src_schema = src.schema();
```

Comment:
`src_schema` will be used multiple times during the iteration so that I think it would be better to save it as a temp variable, in case of compiler is not clever enough. Notice that `src.schema()` might not be a simple function call, because it may become something like `src.src.src.schema()` finally.